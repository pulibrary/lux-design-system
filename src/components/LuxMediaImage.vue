<template>
  <div class="lux-media-image" :class="[height, { 'lux-default-thumbnail': !source }]">
    <img
      v-if="source"
      @error="source = null"
      :src="source"
      :alt="alt"
      :class="[{ 'lux-cover': cover }, { 'lux-contain': contain }]"
    />
    <lux-icon-base
      v-else
      width="50"
      height="50"
      icon-name="file"
      icon-color="rgb(225,225,225)"
      icon-hide="true"
      ><lux-icon-file
    /></lux-icon-base>
  </div>
</template>

<script>
/**
 * Media-Image is a component that is used to display an image,
 * or an icon if the image can't be resolved.
 */
export default {
  name: "LuxMediaImage",
  status: "ready",
  release: "1.0.0",
  type: "Element",
  data: function () {
    return {
      source: this.src,
    }
  },
  props: {
    /**
     * The image displayed
     */
    src: {
      type: String,
      default: null,
    },
    /**
     * The alternative text describing the image. Do not include if image is decorative.
     */
    alt: {
      type: String,
      default: "",
    },
    /**
     * Manually define the height of the image for a card
     */
    height: {
      type: String,
      default: "",
    },
    /**
     * Whether the image fills the container maintaining aspect ratio and is cropped
     */
    cover: {
      type: Boolean,
      default: false,
    },
    /**
     * Whether the full image is contained within the container maintaining aspect ratio. Note that this property is not recommened for use when the `height` prop is used as it will show the background of the container.
     */
    contain: {
      type: Boolean,
      default: false,
    },
  },
}
</script>

<style lang="scss" scoped>
@import "../assets/styles/mixins.scss";
@import "../assets/styles/spacing.scss";

.lux-media-image {
  @include reset;
  @include stack-space(var(--space-base));
  font-family: var(--font-family-text);
  line-height: var(--line-height-base);
  color: var(--color-rich-black);
  overflow: hidden;
  position: relative;
  height: auto;

  img {
    background-color: var(--color-grayscale-lighter);
    height: auto;
    max-width: 100%;
    vertical-align: middle;
  }

  .lux-cover {
    object-fit: cover;
  }

  .lux-contain {
    object-fit: contain;
  }

  .lux-icon {
    display: none;
  }

  &.lux-default-thumbnail {
    background: var(--color-grayscale);
    display: flex;
    align-items: center;

    .lux-icon {
      display: block;
    }
    .lux-svg-icon,
    svg {
      display: block;
      margin: auto;
    }
  }
}

.lux-small {
  height: 100px;
  width: 100%;

  img {
    height: 100px;
    max-width: 100%;
  }
}

.lux-medium {
  height: 200px;
  width: 100%;

  img {
    height: 200px;
    max-width: 100%;
  }
}

.lux-large {
  height: 300px;
  width: 100%;

  img {
    height: 300px;
    max-width: 100%;
  }
}
</style>

<docs>
  ```
  <div>
    <lux-media-image src="https://picsum.photos/400/300/?random" height="medium"></lux-media-image>
  </div>
  ```
</docs>
