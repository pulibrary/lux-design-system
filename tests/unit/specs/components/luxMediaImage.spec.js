import { mount } from "@vue/test-utils"
import LuxMediaImage from "@/components/LuxMediaImage.vue"
import { nextTick } from "vue"

let wrapper

describe("LuxMediaImage.vue", () => {
  beforeEach(() => {
    wrapper = mount(LuxMediaImage, {
      propsData: {
        src: "",
        height: "medium",
        alt: "alt text",
        cover: true,
        contain: false,
      },
      global: {
        stubs: {
          "lux-icon-base": true,
          "lux-icon-file": true,
        },
      },
    })
  })

  it("displays svg icon when src is bad or non-existent", async () => {
    const mediaimg = wrapper.find(".lux-media-image")
    expect(mediaimg.classes()).toContain("medium")
    wrapper.setProps({ height: "large" })
    await nextTick()
    expect(mediaimg.classes()).toContain("large")
    expect(wrapper.find('[icon-name="file"]').exists()).toBe(true)
    expect(wrapper.find("img").exists()).toBe(false)
  })

  it("displays image when src exists", () => {
    const wrapper2 = mount(LuxMediaImage, {
      propsData: {
        src: "https://picsum.photos/400/300/?random",
        height: "medium",
        alt: "alt text",
        cover: false,
        contain: true,
      },
    })
    expect(wrapper2.find('[name="pul-icon-file"]').exists()).toBe(false)
    expect(wrapper2.find("img").exists()).toBe(true)
    const img = wrapper2.find("img")
    expect(img.attributes().alt).toBe("alt text")
    expect(img.classes()).toContain("lux-contain")
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
