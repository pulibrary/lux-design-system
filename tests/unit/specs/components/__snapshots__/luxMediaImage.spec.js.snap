// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuxMediaImage.vue has the expected html structure 1`] = `
<div
  class="lux-media-image medium lux-default-thumbnail"
>
  <lux-icon-base
    height="50"
    icon-color="rgb(225,225,225)"
    icon-hide="true"
    icon-name="file"
    width="50"
  >
    <lux-icon-file />
  </lux-icon-base>
</div>
`;
